server: reduce execution time while listing project if projects have many resource tags

ustcweizhou opened the pull request:
## Description

If projects have many resource tags, it will take long time to list projects.
Remove resource tags information from project_view will fix it the issue.

Fixes: #3178

## Types of changes
<!--- What types of changes does your code introduce? Put an `x` in all the boxes that apply: -->
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] New feature (non-breaking change which adds functionality)
- [X] Bug fix (non-breaking change which fixes an issue)
- [ ] Enhancement (improves an existing feature and functionality)
- [ ] Cleanup (Code refactoring and cleanup, that may add test cases)


## How Has This Been Tested?
<!-- Please describe in detail how you tested your changes. -->
<!-- Include details of your testing environment, and the tests you ran to -->
<!-- see how your change affects other areas of the code, etc. -->

create project,
add resource tags to project,
listprojects  lists all resouce tags.

anuragaw commented:
If the issue is mostly prevalent in UI, I would like to propose an alternate - 
1) Add a parameter ```showTags``` to ```list projects``` API
2) Set the flag to false to list project call in GUI.

Although if not many users need the tags as part of response this is perfectly reasonable too.

svenvogel commented:
@ustcweizhou we will test it and come back :)

ustcweizhou commented:
> If the issue is mostly prevalent in UI, I would like to propose an alternate -
> 
> 1. Add a parameter `showTags` to `list projects` API
> 2. Set the flag to false to list project call in GUI.
> 
> Although if not many users need the tags as part of response this is perfectly reasonable too.

@anuragaw In my point of view,  it would be preferable to fix an issue without api change.

DaanHoogland commented:
not sure if the 'standard' travis smoke tests cover enough of this functionality.
@blueorangutan package

blueorangutan commented:
@DaanHoogland a Jenkins job has been kicked to build packages. I'll keep you posted as I make progress.

blueorangutan commented:
Packaging result: ✔centos6 ✔centos7 ✔debian. JID-2727

DaanHoogland commented:
@blueorangutan test

blueorangutan commented:
@DaanHoogland a Trillian-Jenkins test job (centos7 mgmt + kvm-centos7) has been kicked to run smoke tests

svenvogel commented:
@ustcweizhou we are tested it and it seems and feels faster but it seems there is still a small delay. is there anything to make it more faster?

ustcweizhou commented:
@svenvogel we have similar issue while listing accounts. for project, it can be improved by
```
diff --git a/server/src/main/java/com/cloud/api/query/QueryManagerImpl.java b/server/src/main/java/com/cloud/api/query/QueryManagerImpl.java
index 91e0466..2b15113 100644
--- a/server/src/main/java/com/cloud/api/query/QueryManagerImpl.java
+++ b/server/src/main/java/com/cloud/api/query/QueryManagerImpl.java
@@ -1298,7 +1298,7 @@ public class QueryManagerImpl extends MutualExclusiveIdsManagerBase implements Q
         Long accountId = null;
         String path = null;

-        Filter searchFilter = new Filter(ProjectJoinVO.class, "id", false, startIndex, pageSize);
+        Filter searchFilter = new Filter(ProjectJoinVO.class, null, false, startIndex, pageSize);
         SearchBuilder<ProjectJoinVO> sb = _projectJoinDao.createSearchBuilder();
         sb.select(null, Func.DISTINCT, sb.entity().getId()); // select distinct
         // ids
```

Do not order project_view will reduce execution time of db query.

svenvogel commented:
@ustcweizhou oh i see. hmm should we add this to this pr?

svenvogel commented:
@ustcweizhou we will pick up your change ne try to fix the sort of the project list with jQuery.

DennisKonrad commented:
@ustcweizhou @svenvogel  I built and tested both the PR and the speed improvemennt for the db query (not ordering project_view). Although there seems to be a slight improvement in time for us it still takes way too long. It always feels like you miss-clicked and users then tend to click two or three times (which ends in misery because the lists closes with the second click)

ustcweizhou commented:
@DennisKonrad did you recreate project_view in your testing ?

DennisKonrad commented:
@ustcweizhou Yes, i executed the schema update. I think there is an issue about the project listing that's not about tags used. We have about ~30 projects now and this seems to be too much

ustcweizhou commented:
@DennisKonrad can you login database to see how many projects do you use and how many records in project_view ?
If you do not use tag , the root cause should be different .

DennisKonrad commented:
@ustcweizhou We have ~40 entries and 30 projects that are selectable in the list. Our database is very fast so the queries should not be a problem.

Do you see any improvments that can be done to the project listing code that would make it faster?

ustcweizhou commented:
@DennisKonrad 
I think it might be caused by the queries to get resource limit of project account.
I will have a look.

ustcweizhou commented:
@DennisKonrad I will raise a PR to fix the issue in listProjects and listAccounts.
What cloudstack version do you use ?

DennisKonrad commented:
@ustcweizhou We more or less use the master branch. So you can base the PR on that if you want. Thanks

ustcweizhou commented:
@DennisKonrad PR #3331 to apache/master

rhtyd commented:
@blueorangutan package

blueorangutan commented:
@rhtyd a Jenkins job has been kicked to build packages. I'll keep you posted as I make progress.

blueorangutan commented:
Packaging result: ✔centos6 ✔centos7 ✔debian. JID-2797

rhtyd commented:
@blueorangutan test

blueorangutan commented:
@rhtyd a Trillian-Jenkins test job (centos7 mgmt + kvm-centos7) has been kicked to run smoke tests

rhtyd commented:
@blueorangutan package

blueorangutan commented:
@rhtyd a Jenkins job has been kicked to build packages. I'll keep you posted as I make progress.

blueorangutan commented:
Packaging result: ✔centos6 ✔centos7 ✔debian. JID-2866

PaulAngus commented:
Guys,
personally I suspect this is just hiding the problem, and doesn't help anyone who does want to see tags.

I created the followingissue a few days ago https://github.com/apache/cloudstack/issues/3375
I suspect that this looping through accounts in a project is causing the delay.

??

ustcweizhou commented:
@PaulAngus I think the issue #3375 can be fixed.

in https://github.com/apache/cloudstack/blob/master/server/src/main/java/com/cloud/api/query/ViewResponseHelper.java#L201-L211

with this PR, fillProjectDetails will do nothing, so the tags will be set only once.
```
    public static ProjectResponse fillProjectDetails(ProjectResponse rsp, ProjectJoinVO proj) {
        return s_projectJoinDao.setProjectResponse(rsp, proj);
    }
```

rhtyd commented:
Can you review and test this @anuragaw

rhtyd commented:
Just to confirm this does not break returning of tags in the API response, just the mechanics of how they are found/used @ustcweizhou ? cc @PaulAngus

rhtyd commented:
@blueorangutan package

blueorangutan commented:
@rhtyd a Jenkins job has been kicked to build packages. I'll keep you posted as I make progress.

ustcweizhou commented:
> Just to confirm this does not break returning of tags in the API response, just the mechanics of how they are found/used @ustcweizhou ? cc @PaulAngus

in my testing , tags are still in the api response and same as before.

rhtyd commented:
Thanks @ustcweizhou I'll kick some test and review/lgtm it soon.

blueorangutan commented:
Packaging result: ✔centos6 ✔centos7 ✔debian. JID-52

rhtyd commented:
@blueorangutan test

blueorangutan commented:
@rhtyd a Trillian-Jenkins test job (centos7 mgmt + kvm-centos7) has been kicked to run smoke tests

blueorangutan commented:
<b>Trillian test result (tid-48)</b>
Environment: kvm-centos7 (x2), Advanced Networking with Mgmt server 7
Total time taken: 31665 seconds
Marvin logs: https://github.com/blueorangutan/acs-prs/releases/download/trillian/pr3306-t48-kvm-centos7.zip
Smoke tests completed. 71 look OK, 0 have error(s)
Only failed tests results shown below:


Test | Result | Time (s) | Test File
--- | --- | --- | ---

anuragaw commented on server/src/main/java/com/cloud/api/query/dao/ProjectJoinDaoImpl.java:
```diff
@@ -101,14 +99,6 @@ public ProjectResponse newProjectResponse(ProjectJoinVO proj) {
 
     @Override
     public ProjectResponse setProjectResponse(ProjectResponse rsp, ProjectJoinVO proj) {
```
This method seems to be called in the following path -

`ViewResponseHelper#createProjectResponse() > ApiDBUtils#fillProjectDetails() > ProjectJoinDaoImpl#setProjectResponse` 

I am a little confused by the necessity of this method call at all. Read the below method definitions-


```

ApiDBUtils {
    public static ProjectResponse fillProjectDetails(ProjectResponse rsp, ProjectJoinVO proj) {
        return s_projectJoinDao.setProjectResponse(rsp, proj);
    }
}

ViewResponseHelper {
    public static List<ProjectResponse> createProjectResponse(ProjectJoinVO... projects) {
        Hashtable<Long, ProjectResponse> prjDataList = new Hashtable<Long, ProjectResponse>();
        // Initialise the prjdatalist with the input data
        for (ProjectJoinVO p : projects) {
            ProjectResponse pData = prjDataList.get(p.getId());
            if (pData == null) {
                // first time encountering this vm
                pData = ApiDBUtils.newProjectResponse(p);
            } else {
                // update those  1 to many mapping fields
                pData = ApiDBUtils.fillProjectDetails(pData, p);
            }
            prjDataList.put(p.getId(), pData);
        }
        return new ArrayList<ProjectResponse>(prjDataList.values());
    }
}
```

Why not simply delete this path and modify the ViewResponseHelper to generate the projects response list at one go?

Apologies if I am missing something obvious?

ustcweizhou replied:
@anuragaw 
you are right.
Actually there is similar comment before https://github.com/apache/cloudstack/pull/3306#issuecomment-502142160

anuragaw replied:
Sounds good. I think we should make this simple change then to make code cleaner and easy to comprehend in this PR.

rhtyd commented:
I'll wait for your final change @ustcweizhou, then it it ready for merge. The latest tests LGTM.

ustcweizhou commented:
@rhtyd @anuragaw 
rebased with latest master, and added a separate commit to remove the methods not in use.
If we need to fill more information to project response in the future, we can revert the commit.

rhtyd commented:
Thanks @ustcweizhou
@blueorangutan package

blueorangutan commented:
@rhtyd a Jenkins job has been kicked to build packages. I'll keep you posted as I make progress.

blueorangutan commented:
Packaging result: ✔centos6 ✔centos7 ✔debian. JID-63

rhtyd commented:
@blueorangutan test

blueorangutan commented:
@rhtyd a Trillian-Jenkins test job (centos7 mgmt + kvm-centos7) has been kicked to run smoke tests

blueorangutan commented:
<b>Trillian test result (tid-84)</b>
Environment: kvm-centos7 (x2), Advanced Networking with Mgmt server 7
Total time taken: 17451 seconds
Marvin logs: https://github.com/blueorangutan/acs-prs/releases/download/trillian/pr3306-t84-kvm-centos7.zip
Intermittent failure detected: /marvin/tests/smoke/test_internal_lb.py
Intermittent failure detected: /marvin/tests/smoke/test_public_ip_range.py
Intermittent failure detected: /marvin/tests/smoke/test_reset_vm_on_reboot.py
Intermittent failure detected: /marvin/tests/smoke/test_resource_accounting.py
Intermittent failure detected: /marvin/tests/smoke/test_router_dhcphosts.py
Intermittent failure detected: /marvin/tests/smoke/test_router_dns.py
Intermittent failure detected: /marvin/tests/smoke/test_router_dnsservice.py
Intermittent failure detected: /marvin/tests/smoke/test_routers_iptables_default_policy.py
Intermittent failure detected: /marvin/tests/smoke/test_routers_network_ops.py
Intermittent failure detected: /marvin/tests/smoke/test_routers.py
Intermittent failure detected: /marvin/tests/smoke/test_secondary_storage.py
Intermittent failure detected: /marvin/tests/smoke/test_service_offerings.py
Intermittent failure detected: /marvin/tests/smoke/test_snapshots.py
Intermittent failure detected: /marvin/tests/smoke/test_ssvm.py
Intermittent failure detected: /marvin/tests/smoke/test_templates.py
Intermittent failure detected: /marvin/tests/smoke/test_usage.py
Intermittent failure detected: /marvin/tests/smoke/test_vm_life_cycle.py
Intermittent failure detected: /marvin/tests/smoke/test_vm_snapshots.py
Intermittent failure detected: /marvin/tests/smoke/test_volumes.py
Intermittent failure detected: /marvin/tests/smoke/test_vpc_redundant.py
Intermittent failure detected: /marvin/tests/smoke/test_vpc_router_nics.py
Intermittent failure detected: /marvin/tests/smoke/test_vpc_vpn.py
Smoke tests completed. 51 look OK, 20 have error(s)
Only failed tests results shown below:


Test | Result | Time (s) | Test File
--- | --- | --- | ---
ContextSuite context=TestResetVmOnReboot>:setup | `Error` | 0.00 | test_reset_vm_on_reboot.py
test_01_so_removal_resource_update | `Error` | 1.38 | test_resource_accounting.py
ContextSuite context=TestRouterDHCPHosts>:setup | `Error` | 0.00 | test_router_dhcphosts.py
ContextSuite context=TestRouterDHCPOpts>:setup | `Error` | 0.00 | test_router_dhcphosts.py
ContextSuite context=TestRouterDns>:setup | `Error` | 0.00 | test_router_dns.py
ContextSuite context=TestRouterDnsService>:setup | `Error` | 0.00 | test_router_dnsservice.py
test_02_routervm_iptables_policies | `Error` | 1.31 | test_routers_iptables_default_policy.py
test_01_single_VPC_iptables_policies | `Error` | 4.37 | test_routers_iptables_default_policy.py
test_01_isolate_network_FW_PF_default_routes_egress_true | `Error` | 1.39 | test_routers_network_ops.py
test_02_isolate_network_FW_PF_default_routes_egress_false | `Error` | 1.40 | test_routers_network_ops.py
test_01_RVR_Network_FW_PF_SSH_default_routes_egress_true | `Error` | 6.11 | test_routers_network_ops.py
test_02_RVR_Network_FW_PF_SSH_default_routes_egress_false | `Error` | 5.99 | test_routers_network_ops.py
test_03_RVR_Network_check_router_state | `Error` | 6.00 | test_routers_network_ops.py
ContextSuite context=TestISOUsage>:setup | `Error` | 0.00 | test_usage.py
ContextSuite context=TestLBRuleUsage>:setup | `Error` | 0.00 | test_usage.py
ContextSuite context=TestNatRuleUsage>:setup | `Error` | 0.00 | test_usage.py
ContextSuite context=TestPublicIPUsage>:setup | `Error` | 0.00 | test_usage.py
ContextSuite context=TestSnapshotUsage>:setup | `Error` | 0.00 | test_usage.py
ContextSuite context=TestVmUsage>:setup | `Error` | 0.00 | test_usage.py
ContextSuite context=TestVolumeUsage>:setup | `Error` | 0.00 | test_usage.py
ContextSuite context=TestVpnUsage>:setup | `Error` | 0.00 | test_usage.py
ContextSuite context=TestRouterServices>:setup | `Error` | 0.00 | test_routers.py
test_01_sys_vm_start | `Failure` | 0.11 | test_secondary_storage.py
ContextSuite context=TestCpuCapServiceOfferings>:setup | `Error` | 0.00 | test_service_offerings.py
ContextSuite context=TestServiceOfferings>:setup | `Error` | 2.23 | test_service_offerings.py
ContextSuite context=TestSnapshotRootDisk>:setup | `Error` | 0.00 | test_snapshots.py
test_01_list_sec_storage_vm | `Failure` | 0.04 | test_ssvm.py
test_02_list_cpvm_vm | `Failure` | 0.04 | test_ssvm.py
test_03_ssvm_internals | `Failure` | 0.04 | test_ssvm.py
test_04_cpvm_internals | `Failure` | 0.04 | test_ssvm.py
test_05_stop_ssvm | `Failure` | 0.04 | test_ssvm.py
test_06_stop_cpvm | `Failure` | 0.04 | test_ssvm.py
test_07_reboot_ssvm | `Failure` | 0.04 | test_ssvm.py
test_08_reboot_cpvm | `Failure` | 0.04 | test_ssvm.py
test_09_destroy_ssvm | `Failure` | 0.04 | test_ssvm.py
test_10_destroy_cpvm | `Failure` | 0.04 | test_ssvm.py
test_02_create_template_with_checksum_sha1 | `Error` | 65.48 | test_templates.py
test_03_create_template_with_checksum_sha256 | `Error` | 65.49 | test_templates.py
test_04_create_template_with_checksum_md5 | `Error` | 65.48 | test_templates.py
test_05_create_template_with_no_checksum | `Error` | 65.48 | test_templates.py
test_02_deploy_vm_from_direct_download_template | `Error` | 1.34 | test_templates.py
test_03_deploy_vm_wrong_checksum | `Error` | 1.40 | test_templates.py
ContextSuite context=TestTemplates>:setup | `Error` | 17.69 | test_templates.py
ContextSuite context=Test01DeployVM>:setup | `Error` | 0.00 | test_vm_life_cycle.py
ContextSuite context=Test02VMLifeCycle>:setup | `Error` | 0.00 | test_vm_life_cycle.py
test_14_secure_to_secure_vm_migration | `Error` | 11.45 | test_vm_life_cycle.py
test_15_secured_to_nonsecured_vm_migration | `Error` | 74.40 | test_vm_life_cycle.py
test_16_nonsecured_to_secured_vm_migration | `Error` | 1.28 | test_vm_life_cycle.py
ContextSuite context=TestVmSnapshot>:setup | `Error` | 2.12 | test_vm_snapshots.py
ContextSuite context=TestCreateVolume>:setup | `Error` | 0.00 | test_volumes.py
ContextSuite context=TestVolumes>:setup | `Error` | 0.00 | test_volumes.py
test_01_create_redundant_VPC_2tiers_4VMs_4IPs_4PF_ACL | `Error` | 7.18 | test_vpc_redundant.py
test_02_redundant_VPC_default_routes | `Error` | 8.13 | test_vpc_redundant.py
test_03_create_redundant_VPC_1tier_2VMs_2IPs_2PF_ACL_reboot_routers | `Error` | 6.06 | test_vpc_redundant.py
test_04_rvpc_network_garbage_collector_nics | `Error` | 6.07 | test_vpc_redundant.py
test_05_rvpc_multi_tiers | `Error` | 6.04 | test_vpc_redundant.py
test_01_VPC_nics_after_destroy | `Error` | 3.97 | test_vpc_router_nics.py
test_02_VPC_default_routes | `Error` | 4.00 | test_vpc_router_nics.py
test_01_redundant_vpc_site2site_vpn | `Failure` | 6.50 | test_vpc_vpn.py
test_01_vpc_site2site_vpn_multiple_options | `Failure` | 6.44 | test_vpc_vpn.py
test_01_vpc_remote_access_vpn | `Failure` | 3.23 | test_vpc_vpn.py
test_01_vpc_site2site_vpn | `Failure` | 4.40 | test_vpc_vpn.py

rhtyd commented:
@blueorangutan test

blueorangutan commented:
@rhtyd a Trillian-Jenkins test job (centos7 mgmt + kvm-centos7) has been kicked to run smoke tests

blueorangutan commented:
<b>Trillian test result (tid-86)</b>
Environment: kvm-centos7 (x2), Advanced Networking with Mgmt server 7
Total time taken: 34468 seconds
Marvin logs: https://github.com/blueorangutan/acs-prs/releases/download/trillian/pr3306-t86-kvm-centos7.zip
Intermittent failure detected: /marvin/tests/smoke/test_vpc_redundant.py
Smoke tests completed. 70 look OK, 1 have error(s)
Only failed tests results shown below:


Test | Result | Time (s) | Test File
--- | --- | --- | ---
test_05_rvpc_multi_tiers | `Failure` | 419.22 | test_vpc_redundant.py
test_05_rvpc_multi_tiers | `Error` | 446.49 | test_vpc_redundant.py
